gobblin-config-management/gobblin-config-core/src/main/java/org/apache/gobblin/config/store/zip/IvyConfigStoreFactory.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.gobblin.config.store.hdfs.SimpleHadoopFilesystemConfigStore;
+import org.apache.gobblin.util.DownloadUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URLEncodedUtils;
+
+
+/**
+ * {@link ConfigStoreFactory} that downloads a jar file containing the config store paths through ivy and creates a
+ * {@link ZipFileConfigStore} with it.
+ *
+ * An ivy settings file must be present on the classpath named {@link DownloadUtils#IVY_SETTINGS_FILE_NAME}
+ */
+public class IvyConfigStoreFactory implements ConfigStoreFactory<ZipFileConfigStore> {
+
+  private static final String IVY_SCHEME_PREFIX = "ivy-";
+  private static final String ORG_KEY = "org";
+  private static final String MODULE_KEY = "module";
+  private static final String STORE_PREFIX_KEY = "storePrefix";
+
+  @Override
+  public String getScheme() {
+    return getSchemePrefix() + SimpleHDFSConfigStoreFactory.HDFS_SCHEME_NAME;
+  }
+
+  public String getSchemePrefix() {
+    return IVY_SCHEME_PREFIX;
+  }
+
+  /**
+   * Example configKey URI (configuration is passed as part of the query)
+   *
+   * ivy-hdfs://<hdfsURI>/path/to/config/store?org=<jarOrg>&module=<jarModule>&storePrefix=_CONFIG_STORE
+   *
+   * ivy-hdfs: scheme for this factory
+   * hdfsURI/path/to/config/store: location of HDFS config store (used for getting current version)
+   * org/module: org and module of jar containing config store
+   * storePrefix: prefix to paths in config store
+   */
+  @Override
+  public ZipFileConfigStore createConfigStore(URI configKey) throws ConfigStoreCreationException {
+    if (!configKey.getScheme().equals(getScheme())) {
+      throw new ConfigStoreCreationException(configKey, "Config key URI must have scheme " + getScheme());
+    }
+
+    Properties factoryProps = new Properties();
+    for (NameValuePair param : URLEncodedUtils.parse(configKey, "UTF-8")) {
+      factoryProps.setProperty(param.getName(), param.getValue());
+    }
+
+    String jarOrg = factoryProps.getProperty(ORG_KEY);
+    String jarModule = factoryProps.getProperty(MODULE_KEY);
+
+    if (jarOrg == null || jarModule == null) {
+      throw new ConfigStoreCreationException(configKey, "Config key URI must contain org and module to download from");
+    }
+
+    try {
+      SimpleHDFSStoreMetadata metadata = new SimpleHDFSStoreMetadata(
+          org.apache.hadoop.fs.FileSystem.get(new Configuration()), new Path(configKey.getPath(),
+          SimpleHadoopFilesystemConfigStore.CONFIG_STORE_NAME));
+      String currentVersion = metadata.getCurrentVersion();
+
+      URI[] uris = DownloadUtils.downloadJar(jarOrg, jarModule, currentVersion);
+
+      if (uris.length != 1) {
+        throw new ConfigStoreCreationException(configKey, "Expected one jar file from URI");
+      }
+
+      FileSystem zipFs = FileSystems.newFileSystem(Paths.get(uris[0].getPath()), null);
```

Comment:
Is this zipFs always a local path? If it is, should we use LocalFileSystem? If it is dynamic, then how can we make sure the uris[0]'s fs can be used for other uris?